mmselfsup/models/utils/clip.py

```diff
@@ -0,0 +1,385 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Modified from https://github.com/zejiangh/MILAN
+from collections import OrderedDict
+from typing import Optional, Tuple, Union
+
+import numpy as np
+import torch
+from mmengine.logging import MMLogger
+from torch import nn
+
+
+class LayerNorm(nn.LayerNorm):
+    """Subclass torch's LayerNorm to handle fp16."""
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward function."""
+        orig_type = x.dtype
+        ret = super().forward(x.type(torch.float32))
+        return ret.type(orig_type)
+
+
+class QuickGELU(nn.Module):
+    """A faster version of GELU."""
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward function."""
+        return x * torch.sigmoid(1.702 * x)
+
+
+class ResidualAttentionBlock(nn.Module):
+    """Residual Attention Block (RAB).
+
+    Args:
+        d_model (int): The feature dimension.
+        n_head (int): The number of attention heads.
+        attn_mask (torch.Tensor, optional): The attention mask.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 d_model: int,
+                 n_head: int,
+                 attn_mask: Optional[torch.Tensor] = None,
+                 return_attention: bool = False) -> None:
+        super().__init__()
+
+        self.attn = nn.MultiheadAttention(d_model, n_head)
+        self.ln_1 = LayerNorm(d_model)
+        self.mlp = nn.Sequential(
+            OrderedDict([('c_fc', nn.Linear(d_model, d_model * 4)),
+                         ('gelu', QuickGELU()),
+                         ('c_proj', nn.Linear(d_model * 4, d_model))]))
+        self.ln_2 = LayerNorm(d_model)
+        self.attn_mask = attn_mask
+
+        self.return_attention = return_attention
+
+    def attention(self, x: torch.Tensor) -> torch.Tensor:
+        """Attention function."""
+        self.attn_mask = self.attn_mask.to(
+            dtype=x.dtype,
+            device=x.device) if self.attn_mask is not None else None
+        if self.return_attention:
+            return self.attn(
+                x,
+                x,
+                x,
+                need_weights=self.return_attention,
+                attn_mask=self.attn_mask)
+        else:
+            return self.attn(
+                x,
+                x,
+                x,
+                need_weights=self.return_attention,
+                attn_mask=self.attn_mask)[0]
+
+    def forward(
+        self, x: torch.Tensor
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        """Forward function."""
+        if self.return_attention:
+            x_, attention = self.attention(self.ln_1(x))
+            x = x + x_
+            x = x + self.mlp(self.ln_2(x))
+            return x, attention
+        else:
+            x = x + self.attention(self.ln_1(x))
+            x = x + self.mlp(self.ln_2(x))
+            return x
+
+
+class Transformer(nn.Module):
```

Comment:
The name here is kind of confusing, maybe add more docstring

Reply:
I add docstring for this module, but keeping the name unchanged is better, to keep consistent with that in CLIP.